### PR TITLE
Add faster path for small numbers in .toNumber()

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: lib/bn.js [hunk 1]
```diff
@@ -476,8 +476,14 @@
   };
 
   BN.prototype.toNumber = function toNumber () {
-    assert(this.bitLength() <= 53, 'Number can only safely store up to 53 bits');
-    return parseInt(this.toString(), 10);
+    var length = this.bitLength();
+    assert(length <= 53, 'Number can only safely store up to 53 bits');
+    if (length <= 26) {
+      var ret = this.words[0] & 0x3ffffff;
```
Comment (indutny) [inline lib/bn.js]: This `&` should not be necessary

File: lib/bn.js [hunk 2]
```diff
@@ -476,8 +476,14 @@
...
+      return (this.negative !== 0) ? (ret * -1) : ret;
+    } else {
+      return parseInt(this.toString(), 10);
```
Comment (indutny) [inline lib/bn.js]: I wonder if it should do something like `this.words[1] * 0x4000000 + this.words[0]` here. Should speed up things significantly

Comment (axic) [inline lib/bn.js]: I was thinking about that, but that would mean just limiting toNumber to 52 bits instead of 53 and be done with it.

Comment (indutny) [inline lib/bn.js]: Ah, right! Maybe three cases then?

**Comment by axic** [comment]
@indutny added the fast path for all 3 cases

**Comment by indutny** [comment]
Thanks!

File: lib/bn.js [hunk 3]
```diff
@@ -476,8 +476,19 @@
...
+      ret = this.words[0];
+    } else if (length <= 52) {
+      ret = (this.words[1] * 0x4000000) + this.words[0];
+    } else if (length === 53) {
+      // NOTE: at this stage it is known that the top bit is set
+      ret = 0x10000000000000 + (this.words[1] * 0x4000000) + this.words[0];
+    } else {
+      assert(false, 'Number can only safely store up to 53 bits');
+    }
+    return (this.negative !== 0) ? (ret * -1) : ret;
```
Comment (indutny) [inline lib/bn.js]: Heh, `-ret`, perhaps?

Comment (axic) [inline lib/bn.js]: I guess that works too :)

Done.

**Comment by indutny** [comment]
Landed and released, thank you!

**Comment by axic** [comment]
Thanks!

**Pull request closed without merging**